File: src/preprocess_data.py
from pyspark.ml.feature import StringIndexer, VectorAssembler
from pyspark.sql.functions import col, countDistinct
def preprocess_data(spark, file_path):
    df = spark.read.csv(file_path, inferSchema=True)
    df.printSchema()
    df = df.na.drop()
    selected_features = df.columns
    df = df.select(selected_features).na.drop()
    string_indexers = [StringIndexer(inputCol=column, outputCol=column + "_index", handleInvalid="keep") for column in
                       selected_features if df.select(column).dtypes[0][1] == 'string']
    string_indexed_df = df
    for indexer in string_indexers:
        string_indexed_df = indexer.fit(string_indexed_df).transform(string_indexed_df)

    columns_to_drop = [column for column in selected_features if string_indexed_df.select(column).dtypes[0][1] == 'string']

    df = string_indexed_df.drop(*[col(column) for column in columns_to_drop])
    return df

File: datamart/src/main/scala/preprocess_data.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.ml.feature.{StringIndexer, VectorAssembler}
import org.apache.spark.ml.Pipeline
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.Row
import org.apache.spark.sql.types.StringType
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.linalg.Vectors

object DataPreprocessor {
  def preprocessData(spark: SparkSession, filePath: String): Dataset[Row] = {
    val df = spark.read.option("inferSchema", "true").csv(filePath)
    df.printSchema()

    val selectedFeatures = df.columns
    var tempDF = df.select(selectedFeatures.map(col): _*).na.drop()
    println("this is tempDF.printSchema():")
    tempDF.printSchema()

    // Define string columns
    val stringColumns = selectedFeatures.filter(column => tempDF.schema(column).dataType == StringType)

    // Define VectorAssembler to assemble all features
    val assembler = new VectorAssembler()
                      .setInputCols(tempDF.columns.filterNot(stringColumns.contains))
                      .setOutputCol("features")

    // Create a pipeline with all preprocessing stages
    val pipeline = new Pipeline().setStages(Array(assembler))

    // Fit the pipeline to the data
    val pipelineModel = pipeline.fit(tempDF)

    // Apply the pipeline to transform the data
    val finalDF = pipelineModel.transform(tempDF)

    // Convert the features vector column into an array of doubles
    val toArray = udf((v: Vector) => v.toArray)
    val dfWithArray = finalDF.withColumn("features_array", toArray(col("features")))

    // Explode the array column to create a row for each element in the array
    val explodedDF = dfWithArray.select(explode(col("features_array")).alias("element"))

    explodedDF.printSchema()
    explodedDF
  }
}
